compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -1317,8 +1318,8 @@ pub fn install_ice_hook(
     // by the user. Compiler developers and other rustc users can
     // opt in to less-verbose backtraces by manually setting "RUST_BACKTRACE"
     // (e.g. `RUST_BACKTRACE=1`)
-    if std::env::var_os("RUST_BACKTRACE").is_none() {
-        std::env::set_var("RUST_BACKTRACE", "full");
+    if env::var_os("RUST_BACKTRACE").is_none() {
+        panic::set_backtrace_style(panic::BacktraceStyle::Full);
```

Comment:
This only sets the style, but without the env. variable, [this](https://github.com/rust-lang/rust/blob/b58f647d5488dce73bba517907c44af2c2a618c4/library/std/src/backtrace.rs#L253) will think that backtraces are disabled.

Reply:
Interesting, TIL.

Do you know if that is relevant in this context?

Reply:
Hmm. I thought that it is, otherwise I wouldn't comment, but now that I look at it, maybe it isn't. The [default panic hook](https://github.com/rust-lang/rust/blob/3e633981ef0827bf699559036f1437bca071f804/library/std/src/panicking.rs#L237) does not seem to use `Backtrace::capture` and reads the style directly.

Reply:
Good catch, @Kobzol!

Yes, this is relevant here. We want to enable backtraces (unless there is an explicit setting for it).

Reply:
I'm not sure if the compiler calls `Backtrace::capture` somewhere explicitly, or if it just depends on the default panic hook. If the latter is true, then maybe we really don't need to set the env. variable.

That being said, wrapping this line with `unsafe` seems fine to me, it's not like the compiler uses `#[forbid_unsafe_code)]` anyway, and this line has been clearly working fine in the past.

Reply:
I liked this solution better because it doesn't require me to verify that the compiler only calls this function from single-threaded code. (The usual advantage of of calling safe functions.)

If you prefer to call an unsafe function here, I'd mark the `rustc_driver::install_ice_hook` function `unsafe` and audit all callers to check that they all come from single-threaded code. This is fine, I already did it as part of #124636.

Reply:
I think I like the solution of directly calling `panic::set_backtrace_style(panic::BacktraceStyle::Full);` more. Keeping the invariant that a public function is never called concurrently is quite hard. I think that other places in the compiler that should just use `Backtrace::force_capture` if they don't want to depend on the RUST_BACKTRACE env var. At least what's what I would expect if I wanted to add capturing a backtrace somewhere.

Reply:
We want to make the default hook here always print a backtrace. We can't change the default hook to use `Backtrace::force_capture`. We would have to copy it from libstd into this function instead and make sure to keep it in sync.

Reply:
There are uses of `Backtrace::capture` in compiler/rustc_errors/src/lib.rs and compiler/rustc_log/src/lib.rs. Should those be changed somehow?

Reply:
It depends on whether these call sites want to depend on the RUST_BACKTRACE env var or not. 

For [rustc_log](https://github.com/rust-lang/rust/blob/1d0e4afd4cac09078e12a232508c3e9f8d42535d/compiler/rustc_log/src/lib.rs#L162), it looks like that should indeed use `force_capture`.

Reply:
I've opened https://github.com/rust-lang/rust/pull/125355 for the rustc_log case.

I'm less sure about the compiler/rustc_errors/src/lib.rs cases but I think these should use force_capture too. @tbu-, do you want to open a PR, so we can engage with the diagnostics team?

Thanks for reporting this, @RalfJung!